File: built-in plugins dsp/mb-vocal-eq.cpp
/**
 * MB Vocal EQ
 * Category : effect
 * Type     : vocal
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Vocal-tuned equalizer
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_VOCAL_EQ_H
#define MB_VOCAL_EQ_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbVocalEq : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-vocal-eq";
    static constexpr const char* PLUGIN_NAME    = "MB Vocal EQ";
    static constexpr const char* PLUGIN_TYPE    = "vocal";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float lowCut = 80f;  // range [20, 300]
    float presence = 0f;  // range [-12, 12]
    float air = 0f;  // range [-12, 12]
    };

    MbVocalEq() = default;
    ~MbVocalEq() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.lowCut = std::clamp(params.lowCut, 20f, 300f);
        params.presence = std::clamp(params.presence, -12f, 12f);
        params.air = std::clamp(params.air, -12f, 12f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Vocal EQ
        return input;
    }
};

#endif // MB_VOCAL_EQ_H

File: built-in plugins dsp/mb-vocal-eq.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbVocalEqPlugin: PluginDefinition = { id: 'mb-vocal-eq', slug: 'mb-vocal-eq', name: 'MB Vocal EQ', category: 'effect', type: 'vocal', version: '1.0.0', description: 'Vocal-tuned equalizer', author: 'Max Booster', parameters: [{ id: 'lowCut', name: 'Low Cut', type: 'float', defaultValue: 80, minValue: 20, maxValue: 300, automatable: true }, { id: 'presence', name: 'Presence', type: 'float', defaultValue: 0, minValue: -12, maxValue: 12, automatable: true }, { id: 'air', name: 'Air', type: 'float', defaultValue: 0, minValue: -12, maxValue: 12, automatable: true }], defaultPreset: { lowCut: 80, presence: 0, air: 0 } };

export default MbVocalEqPlugin;
